Add Mutex creation extension methods that take an ACL

carlossanlop opened the pull request:
Approved API Proposal: #41662
Related PR for EventWaitHandle: https://github.com/dotnet/corefx/pull/42213

### Description
We don't currently have a way to create a Mutex with a given ACL in .NET Core. We can modify the ACL, but it would be more secure to have the proper ACL on the object from the start.

### Customer impact
Before this change, customers had to create a `Mutex`, then set its ACLs. This presents a few problems:

- Potential security hole as mutexes can be accessed between creation and modification.
- Porting difficulties as there isn't a 1-1 API replacement
This change addresses those problems by adding a new extension method that allows creating a `Mutex` and ensuring the provided ACLs are set during creation.

If time permits it, we would try to backport this change to 3.1.

carlossanlop commented on src/System.Threading.AccessControl/src/System.Threading.AccessControl.csproj:
```diff
@@ -17,10 +26,15 @@
     <Reference Include="System.Security.Principal.Windows" />
   </ItemGroup>
+  <ItemGroup>
+    <Compile Include="System\Threading\MutexAcl.cs" />
```
Build failing because this file is in the wrong location in the csproj. Fixing...

carlossanlop commented on src/System.Threading.AccessControl/src/System/Threading/MutexAcl.net46.cs:
```diff
@@ -0,0 +1,25 @@
+                throw new ArgumentNullException(nameof(mutexSecurity));
+            }
+
+            return new Mutex(initiallyOwned, name, createdNew, mutexSecurity);
```
This should be `out createdNew`. Fixing...

stephentoub commented on src/System.Threading.AccessControl/src/System/Threading/MutexAcl.cs:
```diff
@@ -0,0 +1,90 @@
+
+                try
+                {
+                    return Mutex.OpenExisting(name);
```
Can we add back the null name support as we did with EventWaitHandle?

stephentoub commented on src/System.Threading.AccessControl/src/System/Threading/MutexAcl.net46.cs:
```diff
@@ -0,0 +1,29 @@
+            if (mutexSecurity == null)
+            {
+                throw new ArgumentNullException(nameof(mutexSecurity));
+            }
```
Why do we need any of this argument validation?  Won't the below Mutex ctor do it all?

stephentoub commented on src/System.Threading.AccessControl/tests/ThreadingAclExtensionsTests.cs:
```diff
@@ -2,13 +2,20 @@
     {
+        #region Test methods
+
+        #region Existence tests
```
FWIW, I find this regions unnecessary clutter.

carlossanlop replied:
I'll eventually merge these unit tests with the ones from EventWaitHandle and Semaphore pretty soon.
Should I split the unit tests in separate files?

stephentoub commented on src/System.Threading.AccessControl/src/System/Threading/MutexAcl.cs:
```diff
@@ -0,0 +1,90 @@
+            if (name.Length > Interop.Kernel32.MAX_PATH)
+            {
+                throw new ArgumentException(SR.Format(SR.Argument_WaitHandleNameTooLong, name), nameof(name));
+            }
```
Same as on the EventWaitHandle changes, can we avoid the path length validation, the name nullness validation, etc.?

carlossanlop commented:
Thanks for your review, @stephentoub . I'll be addressing the comments in this PR as we improve the one for EventHandle. I'll submit a new commit for this soon.

carlossanlop commented on src/System.Threading.AccessControl/tests/MutexAclTests.cs:
```diff
@@ -0,0 +1,148 @@
+﻿using System.Collections.Generic;
```
Missing the license header:
```suggestion
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Collections.Generic;
```

carlossanlop commented on src/System.Threading.AccessControl/tests/AclTests.cs:
```diff
@@ -0,0 +1,10 @@
+﻿namespace System.Threading.Tests
```
Missing the license header:
```suggestion
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

namespace System.Threading.Tests
```

carlossanlop commented on src/System.Threading.AccessControl/src/System/Threading/MutexAcl.cs:
```diff
@@ -0,0 +1,82 @@
+
+                ValidateMutexHandle(handle, name, out createdNew);
+
+                var mutex = new Mutex(initiallyOwned);
```
```suggestion
                Mutex mutex = new Mutex(initiallyOwned);
```

carlossanlop commented on src/System.Threading.AccessControl/src/System/Threading/MutexAcl.cs:
```diff
@@ -0,0 +1,82 @@
+                ValidateMutexHandle(handle, name, out createdNew);
+
+                var mutex = new Mutex(initiallyOwned);
+                var old = mutex.SafeWaitHandle;
```
```suggestion
                SafeWaitHandle old = mutex.SafeWaitHandle;
```

maryamariyan commented:
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your corefx repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/corefx <path to the patch created in step 1>`

danmosemsft commented:
@carlossanlop please make a 3.1 PR with template, you can reuse some of it from #42228. Because just about the only change that could affect existing codepaths is the project file changes, I would describe the risk (to existing functionality) as very low. It's worth a try.

danmosemsft commented:
Actually, scratch that. @ericstj pointed out that we can add this post 3.1 because it's in a separate package and does not need changes to existing netcoreapp code. So we don't need to jam this into 3.1.0.

carlossanlop commented:
> _Actually, scratch that. @ericstj pointed out that we can add this post 3.1 because it's in a separate package and does not need changes to existing netcoreapp code. So we don't need to jam this into 3.1.0._

So should I do anything additional or is this merge enough? @danmosemsft

ericstj commented:
If we take it in servicing, we will need to bump the version and include the package in the build.  Dan are you suggesting we take in servicing, or just let it ship out of master for 5.0 (which can be used downlevel)?

danmosemsft commented:
I figured that this means we don't have to decide yet -- we can see what feedback we get, including from the 1st party converter that originally motivated all this work -- and we have options. (So yes @carlossanlop I suggest to simply get everything into master at this point.)
